File: exercise/java/day07/src/main/java/ci/dependencies/Project.java
package ci.dependencies;

public class Project {
    private final boolean buildsSuccessfully;
    private final TestStatus testStatus;

    private Project(boolean buildsSuccessfully, TestStatus testStatus) {
        this.buildsSuccessfully = buildsSuccessfully;
        this.testStatus = testStatus;
    }

    public static ProjectBuilder builder() {
        return new ProjectBuilder();
    }

    public boolean hasTests() {
        return testStatus != TestStatus.NO_TESTS;
    }

    public String runTests() {
        return testStatus == TestStatus.PASSING_TESTS ? "success" : "failure";
    }

    public String deploy() {
        return buildsSuccessfully ? "success" : "failure";
    }

    public static class ProjectBuilder {
        private boolean buildsSuccessfully;
        private TestStatus testStatus;

        public ProjectBuilder setTestStatus(TestStatus testStatus) {
            this.testStatus = testStatus;
            return this;
        }

        public ProjectBuilder setDeploysSuccessfully(boolean buildsSuccessfully) {
            this.buildsSuccessfully = buildsSuccessfully;
            return this;
        }

        public Project build() {
            return new Project(buildsSuccessfully, testStatus);
        }
    }
}

File: exercise/ts/day07/src/project.ts
import {TestStatus} from "./ci";

export class Project {
    private readonly buildsSuccessfully: boolean;
    private readonly testStatus: TestStatus;

    constructor(buildsSuccessfully: boolean, testStatus: TestStatus) {
        this.buildsSuccessfully = buildsSuccessfully;
        this.testStatus = testStatus;
    }

    static builder(): ProjectBuilder {
        return new ProjectBuilder();
    }

    hasTests(): boolean {
        return this.testStatus !== TestStatus.NoTests;
    }

    runTests(): string {
        return this.testStatus === TestStatus.PassingTests ? 'success' : 'failure';
    }

    deploy(): string {
        return this.buildsSuccessfully ? 'success' : 'failure';
    }
}

export class ProjectBuilder {
    private buildsSuccessfully: boolean;
    private testStatus: TestStatus;

    setTestStatus(testStatus: TestStatus): ProjectBuilder {
        this.testStatus = testStatus;
        return this;
    }

    setDeploysSuccessfully(buildsSuccessfully: boolean): ProjectBuilder {
        this.buildsSuccessfully = buildsSuccessfully;
        return this;
    }

    build(): Project {
        return new Project(this.buildsSuccessfully, this.testStatus);
    }
}